tensornetwork/ncon_interface.py

```diff
@@ -73,31 +74,41 @@ def ncon(tensors: Sequence[Tensor],
   tn, con_edges, out_edges = ncon_network(
       tensors, network_structure, con_order=con_order, out_order=out_order)
 
-  # Contract assuming all edges connecting a given pair of nodes are adjacent
-  # in con_order. If this is not the case, the contraction is sub-optimal
-  # so we throw an exception.
-  prev_nodes = []
   while con_edges:
-    e = con_edges.pop(0)  # pop so that older nodes can be deallocated
-    nodes = e.get_nodes()
-
-    nodes_set = set(nodes)
-    if nodes_set != set(prev_nodes):
-      if not nodes_set.issubset(tn.nodes_set):
-        # the node pair was already contracted
-        raise ValueError("Edge '{}' is not adjacent to other edges connecting "
-                         "'{}' and '{}' in the contraction order.".format(
-                             e, nodes[0], nodes[1]))
-      if not con_edges and len(tn.nodes_set) == 2:
-        # If this already produces the final output, order the edges 
-        # here to avoid transposes in some cases.
-        tn.contract_between(
-            *nodes,
-            name="con({},{})".format(*nodes),
-            output_edge_order=out_edges)
-      else:
-        tn.contract_between(*nodes, name="con({},{})".format(*nodes))
-      prev_nodes = nodes
+    nodes_to_contract = con_edges.pop(0).get_nodes()
+    edges_to_contract = tn.get_shared_edges(*nodes_to_contract)
+
+    # Eat up all parallel edges that are adjacent in the ordering.
+    while con_edges:
```

Comment:
I don't like having this while loop be nested with the same condition as the outer while loop. This seems like an easy way for things to break.

Reply:
Hmm. How do you imagine things breaking? The original version had only one loop, but I felt this obscured what was going on (gathering parallel edges before contracting). I found this way more obvious.

Reply:
Can we do this instead?

```python
for edge in reversed(con_edges):
  if edge in edges_to_contract:
    adjacent_parallel_edges.add(edge)
  else:
    break
con_edges = con_edges[:-len(adjacent_parallel_edges)]
```